src/peers/gov/nasa/jpf/vm/JPF_java_lang_StackStreamFactory.java

```diff
@@ -1,18 +1,211 @@
+  //    and call java method doStackWalk() to apply the stream function.
+  // 4. Depending on the number of stack frames and the need of the
+  //    stream function, doStackWalk() may call the native method
+  //    fetchStackFrames() zero or more times to fetch more stack frames
+  //    to apply the stream function.
+  public static class AbstractStackWalker extends NativePeer {
+
+    // There may be multiple StackWalkers traversing stacks, and the stack
+    // frames are not fetched at once (fetchStackFrames() may be called multiple
+    // times to fetch frames incrementally). This HashMap is used to record
+    // where to continue fetching frames (From a key to the `StackFrame` to
+    // start fetching next time).
+    //
+    // Generation of entry:
+    //     The key is generated at the start of traversal, i.e., when
+    //     callStackWalk() is first invoked indirectly by StackWalker.walk().
+    //     And the info is also stored at this time.
+    // Usage and update:
+    //     The key is passed to fetchStackFrames() as its argument. And the
+    //     stored info is updated each time it fetches more frames.
+    // Deletion:
+    //     The stored info is deleted at the end of traversal, i.e., at return
+    //     of callStackWalk().
+    Map<Integer, StackFrame> nextStartFrames = new HashMap<>();
+
+    private StackFrame getFirstNonStackWalkerFrame(ThreadInfo ti) {
+      StackFrame curFrame = ti.getTopFrame();
+      while (true) {
+        ClassInfo ci = curFrame.getClassInfo();
+        boolean isStackWalker =
+            null != ci.getSuperClass("java.lang.StackWalker");
+        boolean isAbstractStackWalker =
+            null != ci.getSuperClass("java.lang.StackStreamFactory$AbstractStackWalker");
+        if (!isStackWalker && !isAbstractStackWalker) {
+          break;
+        }
+        curFrame = curFrame.getPrevious();
+      }
+      return curFrame;
+    }
+
+    private boolean isBottomFrame(StackFrame frame) {
+      return frame == null;
+    }
+
+    private ElementInfo buildFrameInfoObj(MJIEnv env, StackFrame curFrame) {
+      // Prepare info needed by java.lang.StackFrameInfo
+      int declaringClsObjRef = curFrame.getMethodInfo().getClassInfo().getClassObjectRef();
+      int methodNameRef = env.newString(curFrame.getMethodName());
+      int descriptorRef = env.newString(curFrame.getMethodInfo().getSignature());
+      int bci = curFrame.getPC().getPosition();
+      int fileNameRef = env.newString(curFrame.getSourceFile());
+      int lineNumber = curFrame.getLine();
+      boolean isNative = curFrame.isNative();
+
+      ClassInfo frameInfoCls = env.getSystemClassLoaderInfo().getResolvedClassInfo("java.lang.StackFrameInfo");
+      ElementInfo frameInfoObj = env.getHeap().newObject(frameInfoCls, env.getThreadInfo());
+      frameInfoObj.setReferenceField("declaringClass", declaringClsObjRef);
+      frameInfoObj.setReferenceField("methodName", methodNameRef);
+      // TODO: set methodType
```

Comment:
Would setting the method type involve just the signature or more? Please elaborate a bit.
It may also be a good idea to have a failing test that depends on this functionality so it is obvious that something is not yet implemented.